setup/Core/CompositeTask.cs

```diff
@@ -1,40 +1,52 @@
-﻿using System.Linq;
+﻿using Terraria.ModLoader.Setup.Core.Abstractions;
 
-namespace Terraria.ModLoader.Setup
+namespace Terraria.ModLoader.Setup.Core
 {
 	public class CompositeTask : SetupOperation
 	{
-		private SetupOperation[] tasks;
-		private SetupOperation failed;
+		private readonly SetupOperation[] tasks;
+		private SetupOperation? failed;
 
-		public CompositeTask(ITaskInterface taskInterface, params SetupOperation[] tasks) : base(taskInterface) {
+		public CompositeTask(params SetupOperation[] tasks)
+		{
 			this.tasks = tasks;
 		}
 
-		public override bool ConfigurationDialog() {
-			return tasks.All(task => task.ConfigurationDialog());
+		public override async ValueTask<bool> ConfigurationPrompt(CancellationToken cancellationToken = default)
+		{
+			foreach (var task in tasks) {
+				bool result = await task.ConfigurationPrompt(cancellationToken).ConfigureAwait(false);
+
+				if (!result) {
+					return false;
+				}
+			}
+
+			return true;
 		}
 
-		public override bool Failed() {
+		public override bool Failed()
+		{
 			return failed != null;
 		}
 
-		public override void FinishedDialog() {
+		public override void FinishedPrompt()
+		{
 			if (failed != null)
-				failed.FinishedDialog();
+				failed.FinishedPrompt();
 			else
 				foreach (var task in tasks)
-					task.FinishedDialog();
+					task.FinishedPrompt();
 		}
 
-		public override void Run() {
+		public override async Task Run(IProgress progress, CancellationToken cancellationToken = default)
+		{
 			foreach (var task in tasks) {
-				task.Run();
+				await task.Run(progress, cancellationToken).ConfigureAwait(false);
 				if (task.Failed()) {
 					failed = task;
-					return;
```

Comment:
`return;` removed? Changes semantics. Maybe it never mattered in practice.

See my other notes about replacing the `Failed` checks with exceptions like normal sane people anyway